### ISSUE TITLE
[rush] "Rush install" installs newer versions than specified in pnpm-lock.yaml

### ISSUE THREAD

**Issue by chbota** [microsoft/web-build-tools#1492]
**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**
When running `rush install`, the latest version that adheres to the package.json semver is installed, rather than the version specified in the `pnpm-lock.yaml` file.

I put together a standalone repo that reproduces this behavior here: https://github.com/chbota/rush-pnpm-bug

In this repo, `test-package` depends on `typescript@^3.5.1`. The `pnpm-lock.yaml` file specifies the version as 3.5.1.

After running `rush install`:
- Typescript 3.6.2 (or whatever the currently latest version is that adheres to ^3.5.1 semver) is installed
- common/temp/pnpm-lock.yaml now refers to 3.6.2, while common/config/rush/pnpm-lock.yaml still refers to 3.5.1

If you run "rush update --full", common/config/rush/pnpm-lock.yaml will be updated to refer to 3.6.2. Running just "rush update" appears to have no effect.

**What is the expected behavior?**
Typescript 3.5.1 should be installed (to match the version specified in the git-tracked pnpm-lock.yaml)

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** Rush + pnpm
* **Tool Version:** Rush v5.11.0, pnpm v3.6.0
* **Node Version:** 10.14.1
  * **Is this a LTS version?** yes 
  * **Have you tested on a LTS version?** yes
* **OS:** Windows 10

**Comment by sachinjoseph** [comment]
Hi Chris, looks like your pnpm-lock.yaml was generated when `test-package` depended on `typescript@3.5.1` and then `test-package`'s package.json was updated depend on `typescript@^3.5.1`. With this change, `pnpm-lock.yaml`'s tarball hash for test-package.tgz is invalidated, and now when pnpm install is run (by rush internally), the hash of test-package.tgz and the specifier for TypeScript is updated. 

When this hash is updated, pnpm also seems to update the top level dependency version to the latest `typescript@^3.5.1` - I am not sure if this is a bug or not. As for `test-package`, it still accuartely depends on `typescript@3.5.1` as seen below in the section 'file:projects/test-package.tgz'. The node_modules folder in `test-package` also symlinks to `typescript@3.5.1` even though both the versions are installed. Functionally, this does not look like a bug to me.

```yaml

dependencies:
  '@rush-temp/test-package': 'file:projects/test-package.tgz'
  typescript: 3.6.2
lockfileVersion: 5.1
packages:
  /typescript/3.5.1:
    dev: false
    engines:
      node: '>=4.2.0'
    hasBin: true
    resolution:
      integrity: sha512-64HkdiRv1yYZsSe4xC1WVgamNigVYjlssIoaH2HcZF0+ijsk5YK2g0G34w9wJkze8+5ow4STd22AynfO6ZYYLw==
  /typescript/3.6.2:
    dev: false
    engines:
      node: '>=4.2.0'
    hasBin: true
    resolution:
      integrity: sha512-lmQ4L+J6mnu3xweP8+rOrUwzmN+MRAj7TgtJtDaXE5PMyX2kCrklhg3rvOsOIfNeAWMQWO2F1GPc1kMD2vLAfw==
  'file:projects/test-package.tgz':
    dependencies:
      typescript: 3.5.1
    dev: false
    name: '@rush-temp/test-package'
    resolution:
      integrity: sha512-vCLFkPJhYHKX5v0hKzONlHo8E3tzLf3g8UGd1j/SCmo9N1XttdkZzTU33dp9qDp6f06g5omiDBfAT3NS2CgENw==
      tarball: 'file:projects/test-package.tgz'
    version: 0.0.0
registry: ''
specifiers:
  '@rush-temp/test-package': 'file:./projects/test-package.tgz'
  typescript: ^3.5.1

```

**Comment by chbota** [comment]
Hey Sachin,

Thanks for taking a look.

I think my repro of the issue is incomplete. We're seeing behavior in an internal repository where the version specified in `temp/pnpm-lock.yaml` and the linked version in our package's `node_modules` folder are both `typescript@3.6.2`, rather than `typescript@3.5.1`. In that internal repository, we arrived at this state as a result of the release of `typescript@3.6.2`, and so our package's hash isn't invalid. I was attempting to approximate that repository's state in the repro I've included with this issue. In other words, I'm seeing this behavior with a correct hash in the git-tracked `pnpm-lock.yaml` file, but unfortunately that's in a non-public repository 😄.

In comparing the private repository with the repro above, I realized that we're specifying `shamefully-flatten=true` in `common/config/rush/.npmrc` in our private repository, but not in the repro. I've updated the repro code to include this option, and now the issue (`typescript@3.6.2` being used instead of `typescript@3.5.1` both in `node_modules` and `temp/pnpm-lock.yaml`) is reproducing as described in the original issue. We're using `shamefully-flatten=true` to work around some compatibility issues between `pnpm` and `jest`

**Summary**: the issue appears to only occur when `shamefully-flatten=true`

**Comment by chbota** [comment]
@sachinjoseph , have you had a chance to take another look at this?

**Comment by sachinjoseph** [comment]
No, I haven't had a chance yet to look at this in detail. For now, you should be able specify this version in your preferredVersions:

```json
  "preferredVersions": {
    "typescript": "3.5.1"
  }
```
